File: Programmers/level2/큰 수 만들기.py
def solution(number, k):

    stack = []

    for n in number:
        while k > 0 and stack and stack[-1] < n:
            stack.pop()
            k -= 1
        stack.append(n)

    return ''.join(stack[:len(number) - k])


number = "1231234"
k = 3
print(solution(number, k))

# number = "4177252841"
# k = 4
# print(solution(number, k))

# number = "9876"
# k = 2
# print(solution(number, k))


# 참고:
# https://velog.io/@soo5717/프로그래머스-큰-수-만들기-파이썬

File: Programmers/level2/큰 수 만들기.js
function solution(number, k) {
  const stack = [];

  for (let i = 0; i < number.length; i++) {
    while (k > 0 && stack.length > 0 && stack[stack.length - 1] < number[i]) {
      stack.pop();
      k--;
    }

    stack.push(number[i]);

    if (k === 0) {
      for (let j = i + 1; j < number.length; j++) {
        stack.push(number[j]);
      }
      break;
    }
  }

  const string = stack.join("");

  return string.slice(0, string.length - k);
}

console.log(solution("1924", 2));
console.log(solution("1231234", 3));
console.log(solution("4177252841", 4));
console.log(solution("4321", 2));
